### ISSUE TITLE
Don't let game host see cards before everyone agrees to start

### ISSUE THREAD

**Issue by kberg** [bafolts/terraforming-mars#3198]
As a way of preventing a game host from viewing their hands and the hands of their opponents before sharing a game link, I propose adding this feature: add a game state prior to initial draft in which every player must signal they are ready. Once every player signals they are ready (by clicking on an "I am here" link of sorts) then every player will be eligible to see their starting hands, the map, the MAs, et cetera.

Since each "I am ready" action is single-use, an experienced player will know their game has been tampered with because they haven't been presented with an "I am ready" button.

That display would be really useful as a way for players to see the game parameters, and edit their names, before starting.

**Comment by alrusdi** [comment]
Simple Auth will ultimately solve all the problems like this ;-)

**Comment by alrusdi** [comment]
As an alternative variant we can implement Game Invitation page
1) When host (or Discord bot) hits Create Game - then actual game is not created, but created an Invite (Invitation?) object with game settings and specific link
2) Host (or Discord bot) posts invitation link somewhere (anyone can go to this page)
3) Each Invite has minimal amount of claimable seats. Any person can claim one seat
4) When quorum is reached admin of the game hits Start game. At this point actual game is created from game settings and each person redirected to their player link

As result no one can see player links other than his own. And we add an ability to create game by Discord bot (first participant becomes admin).

**Comment by kevinb9n** [comment]
As a side side benefit I think this extra step would also allow new players to choose for themselves if they want BeginnerCorp, without the game creator having to decide in advance. That is, the option to take a BC for yourself can just always be there, since it doesn't slow anyone down (they have to click something either way).

**Comment by kberg** [comment]
Yeah, what I'm doing at the moment is just setting up the ability for a
player to change their name before starting, as a way to get the "ready"
phase in place. It's interesting.

>

**Comment by ssimeonoff** [comment]
This reminds of a feature I was thinking of : Don't let other players with my link submit actions.

During this initial state, along the button submitting the "ready" state, add an simple input field asking the player to set a password/token if they want to. (leaving the input blank or not)
If the input in not blank, e.g. "123" input has been set.
From the moment on, if there is a POST from this gameID url not including the &t=123 in the url, the server to ignore it and return an error. 
So we can share more freely, as before, the game urls, without worrying that someone will tamper with our games.

**Comment by kberg** [comment]
That's great, because I was thinking about something similar on the mods
channel. Of course, I'd prefer being able to claim a player link through
oauth, but PIN could also work.

>

**Comment by kberg** [comment]
Development update: I got fairly far (You can see the code chagnes in my repo in the branch `ready`) when I realized the server model isn't deal for this. So I'm going to pause and look at breaking up and simplifying the server model, which grew organically.

**Comment by kberg** [comment]
Wow, one year later. Well, the issue has come up again. Server Model is definitely cleaner than before, at least. :D

**Comment by kberg** [comment]
New idea, and here's the full layout:

## User experience
1. Host creates a "table" specifying player count and game options. App returns a URL, let's call it `/table?id=<some unique id>`
2. Host shares the URL with opponents.
3. Each player, including the host, presses enters their name, chooses their color, and presses "Claim".
3a. If player claims a color or name someone else has taken, they may not claim a seat.
3b. If someone claims a seat with the same IP Address as another player, presents a warning. This prevents someone claiming a seat, refreshing the page, and claiming another seat.
4. As players claim seats, the page is updated with the table's state.
5. When all seats are claimed, the service creates a game, and all players get redirected to it.

## Data structures

```
class Table {
  id: string;
  seats: Array<Seat>;
}

class Seat {
  name: string;
  color: Color;
  ip: string;
}
```

This avoids the use case where players can get a beginner corp or handicap. (Beginner corp could be managed by allowing _all_ players to choose the beginner corp as a game option.

## API

`/table/create` (POST GameOptions & player count)
`/table?id=<id>` (GET: View a table and its status.)
`/api/table?id=<id>` (GET: fetch state of table.)
`/table/claim` (POST table id, name, color)
`/admin/tables` (GET view all tables)
